Sources/NIO/SocketOptionChannel.swift

```diff
@@ -0,0 +1,103 @@
+/// options (via `setsockopt`/`getsockopt` or similar) that are larger than
+/// the `SocketOption` `ChannelOption` supports (`Int32`). It provides a strongly
+/// typed API that makes working with larger, less-common socket options easier
+/// than the `ChannelOption` API allows.
+///
+/// Note that while the API has strong types, the API is not itself type-safe.
+/// It is possible to attempt to set a socket option to a value that does not
+/// match the expected type of the option. This will likely fail, and be reported
+/// as an error. This interface is a limitation of the `setsockopt`/`getsockopt`
+/// API.
+///
+/// Not all `Channel` objects support this kind of socket option, so this is not
+/// necessarily widely implemented.
+///
+/// ### Using SocketOptionChannel
+///
+/// To attempt to use these functions, users will need to attempt to conditionally
+/// downcast to the correct protocol type. For example:
+///
+///     ClientBootstrap(group: group).connect(to: target).thenThrowing { channel in
+///         guard let channel = channel as? SocketOptionChannel else {
+///             throw MyDowncastError()
+///         }
+///         return channel
+///     }.then { socketOptionChannel in
+///         socketOptionChannel.setSocketOption(level: SOL_SOCKET, name: SO_REUSEADDR, value: 1)
+///     }
+///
+/// ### Relationship to SocketOption
+///
+/// All `Channel` objects that implement this protocol should also support the
+/// `SocketOption` `ChannelOption` for simple socket options (those with C `int`
+/// values). These are the most common socket option types, and so this `ChannelOption`
+/// represents a convenient shorthand for using this protocol where the type allows.
+///
+/// - note: Like the `Channel` protocol, all methods in this protocol are
+///     thread-safe.
+public protocol SocketOptionChannel: Channel {
+    /// Set a socket option for a given level and name to the specified value.
+    ///
+    /// - parameters:
+    ///     - level: The socket option level, e.g. `SOL_SOCKET` or `IPPROTO_IP`.
+    ///     - name: The name of the socket option, e.g. `SO_REUSEADDR`.
+    ///     - value: The value to set the socket option to.
+    /// - returns: An `EventLoopFuture` that fires when the option has been set,
+    ///     or if an error has occurred.
+    func setSocketOption<Value>(level: SocketOptionLevel, name: SocketOptionName, value: Value) -> EventLoopFuture<Void>
+
+    /// Obtain the value of the socket option for the given level and name.
+    ///
+    /// - parameters:
+    ///     - level: The socket option level, e.g. `SOL_SOCKET` or `IPPROTO_IP`.
+    ///     - name: The name of the socket option, e.g. `SO_REUSEADDR`.
+    /// - returns: An `EventLoopFuture` containing the value of the socket option, or
+    ///     any error that occurred while retrieving the socket option.
+    func getSocketOption<Value>(level: SocketOptionLevel, name: SocketOptionName) -> EventLoopFuture<Value>
```

Comment:
- should we make this have an 'unsafe' or similar word in there?
- other idea would be to make `Value: CSockOptValue` (or similar) and make that a marker protocol which can be assigned to all known types that `set/getsockopt` allow. so `extension timeval: CSockOptValue {} extension CInt: CSockOptValue {}` etc.

Reply:
I just want to prevent users from passing arbitrary stuff (like say an array of bytes) to that `Value` and never ever getting a compiler error. It's really a universal blackhole.